src/main/java/net/minecraftforge/dimension/DynamicDimensionManager.java

```diff
@@ -0,0 +1,221 @@
+import net.minecraft.util.RegistryKey;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.registry.DynamicRegistries;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.Dimension;
+import net.minecraft.world.DimensionType;
+import net.minecraft.world.World;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.gen.ChunkGenerator;
+import net.minecraft.world.gen.DimensionSettings;
+import net.minecraft.world.server.ServerWorld;
+import net.minecraft.world.storage.IServerConfiguration;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.dimension.DimensionRegisterEvent;
+
+public class DynamicDimensionManager 
+{
+	
+	private static DynamicDimensionManager DIM_MANAGER = new DynamicDimensionManager();
+	
+	/*
+	 * MinecraftServer not loaded at time of dimension creation, so we need to hold a copy of the dynamic registries
+	 * Not really pretty but we need it. 
+	 */
+	private DynamicRegistries.Impl dynamicRegistries;
+	private IServerConfiguration serverWorldInfo;
+	
+	public static DynamicDimensionManager getDimensionManager() 
+	{
+		return DIM_MANAGER;
+	}
+	
+	public void initDimensionManager(DynamicRegistries.Impl dynamicregistries, IServerConfiguration serverWorldInfo) 
+	{
+		this.dynamicRegistries = dynamicregistries;
+		this.serverWorldInfo = serverWorldInfo;
+	}
+	
+	public void onIntegratedServerStop()
+	{
+		this.dynamicRegistries = null;
+		this.serverWorldInfo = null;
+	}
+	
+	/**
+	 * Checks if the server is ready to accept registrations of dimension types and dimensions
+	 * @return true if ready, false if not
+	 */
+	public boolean isReady() 
+	{
+		return this.dynamicRegistries != null && this.serverWorldInfo != null;
+	}
+	
+	/**
+	 * INTERNAL METHOD, DO NOT CALL
+	 */
+	public static void onDimensionRegister() 
+	{
+		MinecraftForge.EVENT_BUS.post(new DimensionRegisterEvent(DIM_MANAGER));
+	}
+	
+	/**
+	 * Registers the dimension, this needs to be used one-time at world creation. 
+	 * The dimension will be saved in the level file and loaded when used in the DimensionRegisterEvent.
+	 * This doesn't create the dimension, please refer to {@link loadOrCreateDimension()}
+	 * @param dimension
+	 */
+	public void registerDimension(ResourceLocation loc, Dimension dimension) 
+	{
+		Validate.notNull(loc, "Must provide valid ResourceLocation");
+		Validate.notNull(dimension, "Must provide dimension definition");
+		
+		RegistryKey<Dimension> registryKey = RegistryKey.func_240903_a_(Registry.field_239700_af_, loc);
+		
+		if(this.serverWorldInfo.func_230418_z_().func_236224_e_().func_230516_a_(registryKey) == null) 
+		{
+			this.serverWorldInfo.func_230418_z_().func_236224_e_().register(registryKey, dimension, Lifecycle.stable());
+		}
+	}
+	
+	/**
+	 * Loads or creates the specified dimension.
+	 * @param dimension
+	 */
+	public void loadOrCreateDimension(MinecraftServer server, ResourceLocation loc) 
```

Comment:
you have some weird indentation and spacing going on in this class. Try and keep it consistent

Reply:
That might have been a fuckup of GitHub Desktop. In my file on my local pc, the lines of the method have all the same indentation.